<a href="https://colab.research.google.com/github/CrimsonDynamic/curly-spork/blob/main/Meilenstein_III_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. Einführung
This project investigates Ferrari, McLaren and Williams lap times at the Monza and Monaco circuits from 1996 to 2024. We want to show the evolution of the cars in general and provide a comparison between different teams.

The reason why we chose this time period is, that the governing body of Formula 1, the FiA, introduced a comphrensive and strict rule set regarding the engines in 1996 for the first time in the history of the sport. Before the type of engine (e.g. with turbo or without, what maximum capacity or how many cylinders) weren't regulated.

The primary focus is comparing teams and driver results from 2014 to 2021, the "Hybrid Era". We want to look at the success of the cars over the development cycle, in order to do that, we will compare the laptimes and finishing positions (for each race and for the championship) of the teams.

We will be using the circuits Monza and Monaco to compare data, since both of them have been used throughout 1996 to 2024 without interruption.

The data is sourced from the Kaggle dataset "Formula 1 World Championship 1950-2020" by Rohan Rao and the F1 Wikis of [Ferrari](https://f1.fandom.com/wiki/Scuderia_Ferrari), [McLaren](https://f1.fandom.com/wiki/McLaren), and [Williams](https://f1.fandom.com/wiki/Williams).

# 2. Vergleichbare Datensätze und Analysen
The selected dataset provides detailed information on lap times, drivers, constructors, and races.
Other available datasets often only include race results or summaries but lack detailed lap-by-lap data.
The strength of this dataset lies in its granularity, enabling a deeper analysis.
A somewhat comparable dataset would be [this one](https://www.kaggle.com/datasets/deepshah16/formula-1-19502020?select=AllRace.csv), however it lacks data in lap times, for example.

# 3. Methodik
This project follows the Knowledge Discovery in Databases (KDD) framework:


1. **Selection**: Identify the tables with the necessary information
   - Data Source: Kaggle's "Formula 1 World Championship 1950-2020" dataset, https://f1.fandom.com/wiki/Scuderia_Ferrari.
   - Key Attributes: Lap times, finishing position, and years.


2. **Preprocessing**: Clean the dataset for analysis.
   - Convert lap times to a consistent format (seconds).
   - Filter data for Monza and Monaco
   - Extract Ferrari's, McLaren's and Williams' drivers from 1996 to 2024, as well as the Hybrid era
   - Handle missing values and irrelevant attributes


3. **Transformation**: Prepare the data for specific analyses.
   - Filter for Hybrid era (2014-2021)
   - Group and aggregate data to enable team and driver comparisons


4. **Data Mining**: Analyze trends and patterns.
   - Visualize Ferrari's, McLaren's and Williams' performance at Monza and Monaco (lap times compared to race and years)
   - Visualize the finishing ositions and fastest laps of all constructors to compare them


5. **Interpretation/Evaluation**: Derive insights from findings.
   - Highlight common performance trends over decades
   - Compare competitors performance during the Hybrid era
   - Find out wheter there is a correlation between the fastest lap time and the finishing position


# 4. Data Aquisition
The dataset used in this project is sourced from Kaggle, titled "Formula 1 World Championship 1950-2020" by Rohan Rao (https://www.kaggle.com/datasets/rohanrao/formula-1-world-championship-1950-2020).
It contains detailed information about races, drivers, constructors, lap times, and circuits.

We begin by loading the dataset and inspecting its structure to understand the available attributes and their types.
The focus is on selecting relevant data, particularly the construtors lap times at Monza and Monaco.


In [ ]:
import pandas as pd
import numpy as np

races_url = "https://raw.githubusercontent.com/GrubbySage12/F1-PitstopCrew/refs/heads/main/races.csv"
drivers_url = "https://raw.githubusercontent.com/GrubbySage12/F1-PitstopCrew/refs/heads/main/drivers.csv"
results_url = "https://raw.githubusercontent.com/GrubbySage12/F1-PitstopCrew/refs/heads/main/results.csv"


races_df   = pd.read_csv(races_url, encoding='latin-1')
drivers_df = pd.read_csv(drivers_url, encoding='latin-1')
results_df = pd.read_csv(results_url, encoding='latin-1')

In [ ]:
print("Races DataFrame:")
races_df.head()

Races DataFrame:


raceId  year  round  circuitId                   name        date  \
0       1  2009      1          1  Australian Grand Prix  2009-03-29   
1       2  2009      2          2   Malaysian Grand Prix  2009-04-05   
2       3  2009      3         17     Chinese Grand Prix  2009-04-19   
3       4  2009      4          3     Bahrain Grand Prix  2009-04-26   
4       5  2009      5          4     Spanish Grand Prix  2009-05-10   

       time                                                url fp1_date  \
0  06:00:00  http://en.wikipedia.org/wiki/2009_Australian_G...       \N   
1  09:00:00  http://en.wikipedia.org/wiki/2009_Malaysian_Gr...       \N   
2  07:00:00  http://en.wikipedia.org/wiki/2009_Chinese_Gran...       \N   
3  12:00:00  http://en.wikipedia.org/wiki/2009_Bahrain_Gran...       \N   
4  12:00:00  http://en.wikipedia.org/wiki/2009_Spanish_Gran...       \N   

  fp1_time fp2_date fp2_time fp3_date fp3_time quali_date quali_time  \
0       \N       \N       \N       \N       \N         \N         \N   
1       \N       \N       \N       \N       \N         \N         \N   
2       \N       \N       \N       \N       \N         \N         \N   
3       \N       \N       \N       \N       \N         \N         \N   
4       \N       \N       \N       \N       \N         \N         \N   

  sprint_date sprint_time  
0          \N          \N  
1          \N          \N  
2          \N          \N  
3          \N          \N  
4          \N          \N

In [ ]:
print("\nDrivers DataFrame:")
drivers_df.head()


Drivers DataFrame:


driverId   driverRef number code  forename     surname         dob  \
0         1    hamilton     44  HAM     Lewis    Hamilton  1985-01-07   
1         2    heidfeld     \N  HEI      Nick    Heidfeld  1977-05-10   
2         3     rosberg      6  ROS      Nico     Rosberg  1985-06-27   
3         4      alonso     14  ALO  Fernando      Alonso  1981-07-29   
4         5  kovalainen     \N  KOV    Heikki  Kovalainen  1981-10-19   

  nationality                                             url  
0     British     http://en.wikipedia.org/wiki/Lewis_Hamilton  
1      German      http://en.wikipedia.org/wiki/Nick_Heidfeld  
2      German       http://en.wikipedia.org/wiki/Nico_Rosberg  
3     Spanish    http://en.wikipedia.org/wiki/Fernando_Alonso  
4     Finnish  http://en.wikipedia.org/wiki/Heikki_Kovalainen

In [ ]:
print("\nResults DataFrame:")
results_df.head()


Results DataFrame:


resultId  raceId  driverId  constructorId number  grid position  \
0         1      18         1              1     22     1        1   
1         2      18         2              2      3     5        2   
2         3      18         3              3      7     7        3   
3         4      18         4              4      5    11        4   
4         5      18         5              1     23     3        5   

  positionText  positionOrder  points  laps         time milliseconds  \
0            1              1    10.0    58  1:34:50.616      5690616   
1            2              2     8.0    58       +5.478      5696094   
2            3              3     6.0    58       +8.163      5698779   
3            4              4     5.0    58      +17.181      5707797   
4            5              5     4.0    58      +18.014      5708630   

  fastestLap rank fastestLapTime fastestLapSpeed  statusId  
0         39    2       1:27.452         218.300         1  
1         41    3       1:27.739         217.586         1  
2         41    5       1:28.090         216.719         1  
3         58    7       1:28.603         215.464         1  
4         43    1       1:27.418         218.385         1

## Webscraping

This section was created with the help of Chat GPT, since we only discussed surface details of webscraping in our lectures

We decided to webscrape the drivers for the different eras and teams, in order to simplify merging data sets later.

### Ferrari

In [ ]:
import requests
import urllib.request
from bs4 import BeautifulSoup

my_urlf = "https://f1.fandom.com/wiki/Scuderia_Ferrari#Season-by-season_record"
response = requests.get(my_urlf)

headers = {'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:63.0) Gecko/20100101 Firefox/63.0'}

req = urllib.request.Request(url=my_urlf, headers=headers)

with urllib.request.urlopen(req) as response:
  soup = BeautifulSoup(response.read(), 'html.parser')


record_section = soup.find('span', id='Season-by-season_record')

if record_section:
  parent = record_section.find_parent()
  tables = parent.find_all_next('table')
  if len(tables) > 1:
    table = tables[1]
    driver_listf = set()

    for row in table.find_all('tr')[1:]:
      columns = row.find_all('td')

      year_text = columns[0].get_text(strip=True)
      year = int(year_text)

      if 1996 <= year <= 2024:
        drivers_td = columns[4]
        driversf = [a.get_text(strip=True) for a in drivers_td.find_all('a')]
        driver_listf.update(driversf)
else:
        print("Abschnitt 'Season-by-season record' wurde nicht gefunden.")


for name in sorted(driver_listf):
    print(name)

df_ferrari_complete = pd.DataFrame(driver_listf, columns=['Driver'])
df_ferrari_complete.to_csv('ferrari_drivers.csv', index=False)

Carlos Sainz, Jr.
Charles Leclerc
Eddie Irvine
Felipe Massa
Fernando Alonso
Giancarlo Fisichella
Kimi Räikkönen
Luca Badoer
Michael Schumacher
Mika Salo
Oliver Bearman
Rubens Barrichello
Sebastian Vettel


In [ ]:
import requests
import urllib.request
from bs4 import BeautifulSoup

my_urlf = "https://f1.fandom.com/wiki/Scuderia_Ferrari#Season-by-season_record"
response = requests.get(my_urlf)

headers = {'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:63.0) Gecko/20100101 Firefox/63.0'}

req = urllib.request.Request(url=my_urlf, headers=headers)

with urllib.request.urlopen(req) as response:
  soup = BeautifulSoup(response.read(), 'html.parser')


record_section = soup.find('span', id='Season-by-season_record')

if record_section:
  parent = record_section.find_parent()
  tables = parent.find_all_next('table')
  if len(tables) > 1:
    table = tables[1]
    driver_listf = set()

    for row in table.find_all('tr')[1:]:
      columns = row.find_all('td')

      year_text = columns[0].get_text(strip=True)
      year = int(year_text)

      if 2014 <= year <= 2021:
        drivers_td = columns[4]
        driversf = [a.get_text(strip=True) for a in drivers_td.find_all('a')]
        driver_listf.update(driversf)
else:
        print("Abschnitt 'Season-by-season record' wurde nicht gefunden.")


for name in sorted(driver_listf):
    print(name)

df_ferrari_hybrid = pd.DataFrame(driver_listf, columns=['Driver'])
df_ferrari_hybrid.to_csv('ferrari_drivers_hybrid_era.csv', index=False)

Carlos Sainz, Jr.
Charles Leclerc
Fernando Alonso
Kimi Räikkönen
Sebastian Vettel


### McLaren

In [ ]:
import requests
from bs4 import BeautifulSoup

url = "https://f1.fandom.com/wiki/McLaren"
response = requests.get(url)
soup = BeautifulSoup(response.text, 'html.parser')

season_table = soup.find('span', {'id': 'Year-by-year'}).find_next('table', {'class': 'wikitable'})
rows = season_table.find_all('tr')
drivers_set = set()

for row in rows:
    cols = row.find_all('td')
    if len(cols) > 5:
      year = int(cols[0].text.strip())
      if 1996 <= year <= 2024:
        driver_cell = cols[5]
        drivers = {link.text.strip() for link in driver_cell.find_all('a') if link.text.strip()}
        drivers_set.update(drivers)

for driver in sorted(drivers_set):
    print(driver)

df_mclaren_complete = pd.DataFrame(drivers_set, columns=['Driver'])
df_mclaren_complete.to_csv('mclaren_drivers.csv', index=False)

Alexander Wurz
Carlos Sainz, Jr.
Daniel Ricciardo
David Coulthard
Fernando Alonso
Heikki Kovalainen
Jenson Button
Juan Pablo Montoya
Kevin Magnussen
Kimi Räikkönen
Lando Norris
Lewis Hamilton
Mika Häkkinen
Oscar Piastri
Pedro de la Rosa
Sergio Pérez
Stoffel Vandoorne


In [ ]:
import requests
from bs4 import BeautifulSoup

url = "https://f1.fandom.com/wiki/McLaren"
response = requests.get(url)
soup = BeautifulSoup(response.text, 'html.parser')

season_table = soup.find('span', {'id': 'Year-by-year'}).find_next('table', {'class': 'wikitable'})
rows = season_table.find_all('tr')
drivers_set = set()

for row in rows:
    cols = row.find_all('td')
    if len(cols) > 5:
      year = int(cols[0].text.strip())
      if 2014 <= year <= 2021:
        driver_cell = cols[5]
        drivers = {link.text.strip() for link in driver_cell.find_all('a') if link.text.strip()}
        drivers_set.update(drivers)

for driver in sorted(drivers_set):
    print(driver)

df_mclaren_hybrid = pd.DataFrame(drivers_set, columns=['Driver'])
df_mclaren_hybrid.to_csv('mclaren_drivers_hybrid_era.csv', index=False)

Carlos Sainz, Jr.
Daniel Ricciardo
Fernando Alonso
Jenson Button
Kevin Magnussen
Lando Norris
Stoffel Vandoorne


### Williams

In [ ]:
import requests
from bs4 import BeautifulSoup

url = "https://f1.fandom.com/wiki/Williams"
response = requests.get(url)
soup = BeautifulSoup(response.text, 'html.parser')
season_heading = soup.find('span', {'id': 'Season-by-Season_Record'})
season_table = season_heading.find_next('table', {'class': 'wikitable'})
rows = season_table.find_all('tr')

drivers_list = set()

for row in rows:
    cols = row.find_all('td')
    if len(cols) > 3:
        year_cell = row.find('th')  # Die Jahreszahl ist in der ersten Spalte (th)
        if year_cell:
          year = int(year_cell.text.strip())
          if 1996 <= year <= 2024:
            driver_cell = cols[3]
            driver_links = driver_cell.find_all('a')
            drivers = [link.text.strip() for link in driver_links if link.text.strip()]
            drivers_list.update(drivers)

for driver in sorted(drivers_list):
    print(driver)

df_williams_complete = pd.DataFrame(drivers_list, columns=['Driver'])
df_williams_complete.to_csv('williams_drivers.csv', index=False)

Alex Albon
Alexander Albon
Damon Hill
Felipe Massa
George Russell
Jacques Villeneuve
Juan Pablo Montoya
Lance Stroll
Mark Webber
Nicholas Latifi
Nico Rosberg
Pastor Maldonado
Ralf Schumacher
Rubens Barrichello


In [ ]:
import requests
from bs4 import BeautifulSoup

url = "https://f1.fandom.com/wiki/Williams"
response = requests.get(url)
soup = BeautifulSoup(response.text, 'html.parser')
season_heading = soup.find('span', {'id': 'Season-by-Season_Record'})
season_table = season_heading.find_next('table', {'class': 'wikitable'})
rows = season_table.find_all('tr')

drivers_list = set()

for row in rows:
    cols = row.find_all('td')
    if len(cols) > 3:
        year_cell = row.find('th')  # Die Jahreszahl ist in der ersten Spalte (th)
        if year_cell:
          year = int(year_cell.text.strip())
          if 2014 <= year <= 2021:
            driver_cell = cols[3]
            driver_links = driver_cell.find_all('a')
            drivers = [link.text.strip() for link in driver_links if link.text.strip()]
            drivers_list.update(drivers)

for driver in sorted(drivers_list):
    print(driver)

df_williams_hybrid = pd.DataFrame(drivers_list, columns=['Driver'])
df_williams_hybrid.to_csv('williams_drivers_hybrid_era.csv', index=False)

Felipe Massa
George Russell
Lance Stroll


# 5. Initial Data Analysis / Data Summary / ggf. erste Meta-Analysen

### Races Dataframe

In [ ]:
races_df.describe()

raceId         year        round    circuitId
count  1125.000000  1125.000000  1125.000000  1125.000000
mean    565.710222  1992.703111     8.579556    23.889778
std     328.813817    20.603848     5.159910    19.633527
min       1.000000  1950.000000     1.000000     1.000000
25%     282.000000  1977.000000     4.000000     9.000000
50%     563.000000  1994.000000     8.000000    18.000000
75%     845.000000  2011.000000    13.000000    34.000000
max    1144.000000  2024.000000    24.000000    80.000000

In [ ]:
races_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1125 entries, 0 to 1124
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   raceId       1125 non-null   int64 
 1   year         1125 non-null   int64 
 2   round        1125 non-null   int64 
 3   circuitId    1125 non-null   int64 
 4   name         1125 non-null   object
 5   date         1125 non-null   object
 6   time         1125 non-null   object
 7   url          1125 non-null   object
 8   fp1_date     1125 non-null   object
 9   fp1_time     1125 non-null   object
 10  fp2_date     1125 non-null   object
 11  fp2_time     1125 non-null   object
 12  fp3_date     1125 non-null   object
 13  fp3_time     1125 non-null   object
 14  quali_date   1125 non-null   object
 15  quali_time   1125 non-null   object
 16  sprint_date  1125 non-null   object
 17  sprint_time  1125 non-null   object
dtypes: int64(4), object(14)
memory usage: 158.3+ KB


### Drivers Dataframe

In [ ]:
drivers_df.describe()

driverId
count  859.000000
mean   430.059371
std    248.213115
min      1.000000
25%    215.500000
50%    430.000000
75%    644.500000
max    860.000000

In [ ]:
drivers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 859 entries, 0 to 858
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   driverId     859 non-null    int64 
 1   driverRef    859 non-null    object
 2   number       859 non-null    object
 3   code         859 non-null    object
 4   forename     859 non-null    object
 5   surname      859 non-null    object
 6   dob          859 non-null    object
 7   nationality  859 non-null    object
 8   url          859 non-null    object
dtypes: int64(1), object(8)
memory usage: 60.5+ KB


### Results Dataframe

In [ ]:
results_df.describe()

resultId        raceId      driverId  constructorId          grid  \
count  26519.000000  26519.000000  26519.000000   26519.000000  26519.000000   
mean   13260.940986    546.376560    274.357291      49.801161     11.145820   
std     7656.813206    309.642244    279.275606      61.091426      7.213453   
min        1.000000      1.000000      1.000000       1.000000      0.000000   
25%     6630.500000    298.000000     57.000000       6.000000      5.000000   
50%    13260.000000    527.000000    170.000000      25.000000     11.000000   
75%    19889.500000    803.000000    385.000000      60.000000     17.000000   
max    26524.000000   1132.000000    860.000000     215.000000     34.000000   

       positionOrder        points          laps      statusId  
count   26519.000000  26519.000000  26519.000000  26519.000000  
mean       12.814812      1.959578     46.228251     17.317056  
std         7.677869      4.306475     29.577860     26.081700  
min         1.000000      0.000000      0.000000      1.000000  
25%         6.000000      0.000000     23.000000      1.000000  
50%        12.000000      0.000000     53.000000     10.000000  
75%        18.000000      2.000000     66.000000     14.000000  
max        39.000000     50.000000    200.000000    141.000000

In [ ]:
results_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26519 entries, 0 to 26518
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   resultId         26519 non-null  int64  
 1   raceId           26519 non-null  int64  
 2   driverId         26519 non-null  int64  
 3   constructorId    26519 non-null  int64  
 4   number           26519 non-null  object 
 5   grid             26519 non-null  int64  
 6   position         26519 non-null  object 
 7   positionText     26519 non-null  object 
 8   positionOrder    26519 non-null  int64  
 9   points           26519 non-null  float64
 10  laps             26519 non-null  int64  
 11  time             26519 non-null  object 
 12  milliseconds     26519 non-null  object 
 13  fastestLap       26519 non-null  object 
 14  rank             26519 non-null  object 
 15  fastestLapTime   26519 non-null  object 
 16  fastestLapSpeed  26519 non-null  object 
 17  statusId    

### Result

The initial data analysis shows that the "results_df" dataset contains 26,519 entries with 18 columns. The data includes information on race results, such as positions, laps, fastest laps, points, and times. Certain columns, such as "milliseconds" and "fastestLapSpeed", are stored as strings and may require conversion for further analysis.

The "drivers_df" dataset comprises 859 entries with 9 columns, providing details on drivers, including their names, nationalities, and dates of birth. The dataset is complete and does not contain missing values.

The "races_df" dataset includes 1,125 entries with 18 columns, capturing race details such as year, round, circuit, and session timings. The races span from 1950 to 2024, with all values present.

Overall, the datasets provide a good source of information for examining Formula 1 race history, drivers, and performance metrics. Some columns may require preprocessing to facilitate further analysis.


# 6. Erste Vorschläge/ Beschreibung / Pseudo-Code für Data Preprocessing


**Pseudo-Code:**


z.B. Handling Missing Values (Null values):
  
- df.replace({ "/N": "Unknown"}, inplace=True)
- df['milliseconds'].fillna(value='Unknown', inplace=True)  # Replace missing 'milliseconds' with "Unknown"

- df.dropna(subset=['milliseconds'], inplace=True) #Drop rows where any of the relevant columns

# 7. Herausforderungen für weiteren Projektverlauf.

**Feature Selection**: Because we have multiple DataFrames and numerous features, we will select only the features that are most relevant for our visualization. This ensures a clear and concise analysis, focusing on metrics like average lap times, fastest laps, and Ferrari-specific performance across years.



**Handling Missing Values (Null values)**: Check for Null values: We need to identify any missing or null values in our dataset, especially in critical columns such as "milliseconds", "fastestLapTime", and "position", which are essential for our analysis and visualization.
it can be done with

- **1. Imputation:** For columns like "milliseconds" or "fastestLapTime" where the missing data is sparse, we may decide to fill these missing values using imputation methods such as filling with "unknown" value for that column.

- **2. Removal of Rows:** In cases where rows contain a large number of missing values (especially in the "position" column, which is essential for race standings), we may choose to remove these rows to maintain data integrity for our analysis.


# 8. Data Preprocessing

In [ ]:
import numpy as np
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

In [ ]:
races_url = "https://raw.githubusercontent.com/GrubbySage12/F1-PitstopCrew/refs/heads/main/races.csv"
drivers_url = "https://raw.githubusercontent.com/GrubbySage12/F1-PitstopCrew/refs/heads/main/drivers.csv"
results_url = "https://raw.githubusercontent.com/GrubbySage12/F1-PitstopCrew/refs/heads/main/results.csv"
laptimes_url = "https://raw.githubusercontent.com/GrubbySage12/F1-PitstopCrew/refs/heads/main/lap_times.csv"
constructors_url = 'https://raw.githubusercontent.com/GrubbySage12/F1-PitstopCrew/refs/heads/main/constructors.csv'

races_df   = pd.read_csv(races_url, encoding='utf-8')
drivers_df = pd.read_csv(drivers_url, encoding='utf-8')
results_df = pd.read_csv(results_url, encoding='utf-8')
laptimes_df = pd.read_csv(laptimes_url, encoding='utf-8')
constructors_df = pd.read_csv(constructors_url, encoding='utf-8')

In [ ]:
df_ferrari_drivers = df_ferrari_complete
df_ferrari_hybrid_drivers = df_ferrari_hybrid
df_mclaren_drivers = df_mclaren_complete
df_mclaren_hybrid_drivers = df_mclaren_hybrid
df_williams_drivers = df_williams_complete
df_williams_hybrid_drivers = df_williams_hybrid

## Extracting drivers and lap times

Firstly, we need to filter for the races, that are relevant to our analysis. Then we need to add the information of the drivers, carring over relevant attributes. Then, we can sort them.

In [ ]:
filtered_races = races_df[(races_df['year'] >= 1996) & (races_df['year'] <= 2024)]

lap_times_with_year = laptimes_df.merge(filtered_races, on='raceId', how='inner')

In [ ]:
lap_times_with_drivers = lap_times_with_year.merge(drivers_df, on='driverId', how='inner')

merged_df = lap_times_with_drivers[['driverId','forename', 'surname', 'raceId','year', 'time_x','lap']]

lap_times_with_drivers_and_results = merged_df.merge(results_df, on=['raceId', 'driverId'], how='inner')

In [ ]:
final_df = lap_times_with_drivers_and_results.merge(constructors_df, on='constructorId', how='inner')
final_df = final_df[['driverId', 'forename', 'surname', 'raceId', 'year','lap', 'time_x', 'name', 'fastestLapTime']]
final_df = final_df.sort_values(by=['year', 'lap', 'raceId'])
#final_df.to_csv("lap_times_with_drivers_and_constructors_1990_2024.csv", index=False)

In [ ]:
final_df.reset_index(drop=True, inplace=True)
final_df

driverId forename     surname  raceId  year  lap    time_x      name  \
0             30  Michael  Schumacher     224  1996    1  1:45.188   Ferrari   
1             56    Eddie      Irvine     224  1996    1  1:44.981   Ferrari   
2             55     Jean       Alesi     224  1996    1  1:46.506  Benetton   
3             77  Gerhard      Berger     224  1996    1  1:49.726  Benetton   
4             71    Damon        Hill     224  1996    1  1:44.243  Williams   
...          ...      ...         ...     ...   ...  ...       ...       ...   
575024       832   Carlos       Sainz    1128  2024   78  1:17.072   Ferrari   
575025       846    Lando      Norris    1128  2024   78  1:17.296   McLaren   
575026       847   George     Russell    1128  2024   78  1:16.303  Mercedes   
575027       830      Max  Verstappen    1128  2024   78  1:16.248  Red Bull   
575028         1    Lewis    Hamilton    1128  2024   78  1:16.355  Mercedes   

       fastestLapTime  
0                  \N  
1                  \N  
2                  \N  
3                  \N  
4                  \N  
...               ...  
575024             \N  
575025             \N  
575026             \N  
575027             \N  
575028             \N  

[575029 rows x 9 columns]

## Extracting finishing positions

Now, we can also add the finishing positions.

In [ ]:
race_results = results_df.merge(filtered_races, on='raceId', how='inner')

race_results_with_drivers = race_results.merge(drivers_df, on='driverId', how='inner')

race_results_with_positions = race_results_with_drivers[['driverId', 'forename', 'surname', 'raceId', 'year', 'positionOrder']]
race_results_with_positions = race_results_with_positions.sort_values(by=['year', 'raceId'])

In [ ]:
print(race_results_with_positions.sort_values(by=['year', 'raceId', 'positionOrder']))

       driverId forename     surname  raceId  year  positionOrder
4380         71    Damon        Hill     224  1996              1
4381         35  Jacques  Villeneuve     224  1996              2
4382         56    Eddie      Irvine     224  1996              3
4383         77  Gerhard      Berger     224  1996              4
4384         57     Mika    Häkkinen     224  1996              5
...         ...      ...         ...     ...   ...            ...
11251       839  Esteban        Ocon    1132  2024             16
11252       815   Sergio       Pérez    1132  2024             17
11253       855   Guanyu        Zhou    1132  2024             18
11254       847   George     Russell    1132  2024             19
11255       842   Pierre       Gasly    1132  2024             20

[11256 rows x 6 columns]


## Merge dataframes

Lastly, we need the fastest lap time for each race.

In [ ]:
df_merged = final_df.merge(race_results_with_positions, on=['raceId', 'driverId', 'forename', 'surname', 'year'], how='right')

df_final = pd.DataFrame(data=df_merged, columns=['driverId', 'forename', 'surname', 'raceId', 'year', 'lap', 'time_x', 'name', 'fastestLapTime', 'positionOrder'])

In [ ]:
df_final

driverId forename  surname  raceId  year  lap    time_x      name  \
0             71    Damon     Hill     224  1996  1.0  1:44.243  Williams   
1             71    Damon     Hill     224  1996  2.0  1:37.214  Williams   
2             71    Damon     Hill     224  1996  3.0  1:36.249  Williams   
3             71    Damon     Hill     224  1996  4.0  1:35.687  Williams   
4             71    Damon     Hill     224  1996  5.0  1:35.302  Williams   
...          ...      ...      ...     ...   ...  ...       ...       ...   
575491       839  Esteban     Ocon    1132  2024  NaN       NaN       NaN   
575492       815   Sergio    Pérez    1132  2024  NaN       NaN       NaN   
575493       855   Guanyu     Zhou    1132  2024  NaN       NaN       NaN   
575494       847   George  Russell    1132  2024  NaN       NaN       NaN   
575495       842   Pierre    Gasly    1132  2024  NaN       NaN       NaN   

       fastestLapTime  positionOrder  
0                  \N              1  
1                  \N              1  
2                  \N              1  
3                  \N              1  
4                  \N              1  
...               ...            ...  
575491            NaN             16  
575492            NaN             17  
575493            NaN             18  
575494            NaN             19  
575495            NaN             20  

[575496 rows x 10 columns]

## Extracting race IDs for monaco and monza

We also need the raceIds for the Monaco and Monza races.

In [ ]:
mask_monaco = filtered_races['name'] == 'Monaco Grand Prix'
mask_monza = filtered_races['name'] == 'Italian Grand Prix'

df_monaco = pd.DataFrame(data=filtered_races[mask_monaco], columns=['year', 'raceId'])
df_monza = pd.DataFrame(data=filtered_races[mask_monza], columns=['year', 'raceId'])

In [ ]:
df_monaco.reset_index(drop=True, inplace=True)
df_monza.reset_index(drop=True, inplace=True)

df_monaco.sort_values(by='year', inplace=True)
df_monza.sort_values(by='year', inplace=True)

In [ ]:
df_monza.head()

year  raceId
13  1996     237
12  1997     219
11  1998     204
10  1999     187
9   2000     171

In [ ]:
df_monaco.head()

year  raceId
13  1996     229
12  1997     211
11  1998     196
10  1999     178
9   2000     164

## Data Cleaning

### Fill missing values

In order to fill the missing fastest lap times, we use the fastest lap out of all recorded laps for each race (time_x).

In [ ]:
fastestlaptimes_final_df = df_final[['driverId', 'forename', 'surname', 'year',  'name', 'raceId', 'fastestLapTime', 'positionOrder']]
fastestlaptimes_final_df = fastestlaptimes_final_df.drop_duplicates(subset=['raceId', 'driverId'], keep='first')
fastestlaptimes_final_df = fastestlaptimes_final_df[['driverId', 'forename', 'surname', 'year', 'name', 'raceId', 'fastestLapTime', 'positionOrder']]

In [ ]:
def fill_fastest_lap(row):
    if row["fastestLapTime"] == "\\N" or pd.isna(row["fastestLapTime"]):
        times_in_race = final_df[
            (final_df["driverId"] == row["driverId"]) & (final_df["raceId"] == row["raceId"])
        ]["time_x"]
        return min(times_in_race) if not times_in_race.empty else "\\N"
    return row["fastestLapTime"]

In [ ]:
fastestlaptimes_final_df["fastestLapTime"] = fastestlaptimes_final_df.apply(fill_fastest_lap, axis=1)

print(fastestlaptimes_final_df)

        driverId forename     surname  year      name  raceId fastestLapTime  \
0             71    Damon        Hill  1996  Williams     224       1:33.621   
58            35  Jacques  Villeneuve  1996  Williams     224       1:33.421   
116           56    Eddie      Irvine  1996   Ferrari     224       1:34.533   
174           77  Gerhard      Berger  1996  Benetton     224       1:34.757   
232           57     Mika    Häkkinen  1996   McLaren     224       1:35.843   
...          ...      ...         ...   ...       ...     ...            ...   
575491       839  Esteban        Ocon  2024       NaN    1132             \N   
575492       815   Sergio       Pérez  2024       NaN    1132             \N   
575493       855   Guanyu        Zhou  2024       NaN    1132             \N   
575494       847   George     Russell  2024       NaN    1132             \N   
575495       842   Pierre       Gasly  2024       NaN    1132             \N   

        positionOrder  
0              

### Convert lap times

In order to visulise the lap times, we need a consitent format. We agreed on milliseconds.

In [ ]:
def convert_to_milliseconds(time_str):
    if time_str == "\\N" or pd.isna(time_str):
        return np.nan
    try:
        minutes, rest = time_str.split(":")
        seconds, milliseconds = rest.split(".")
        total_ms = (
            int(minutes) * 60 * 1000 +
            int(seconds) * 1000 +
            int(milliseconds)
        )
        return total_ms

    except ValueError:
        return np.nan

In [ ]:
fastestlaptimes_final_df["fastestLapTime_ms"] = fastestlaptimes_final_df["fastestLapTime"].apply(convert_to_milliseconds)

print(fastestlaptimes_final_df)

        driverId forename     surname  year      name  raceId fastestLapTime  \
0             71    Damon        Hill  1996  Williams     224       1:33.621   
58            35  Jacques  Villeneuve  1996  Williams     224       1:33.421   
116           56    Eddie      Irvine  1996   Ferrari     224       1:34.533   
174           77  Gerhard      Berger  1996  Benetton     224       1:34.757   
232           57     Mika    Häkkinen  1996   McLaren     224       1:35.843   
...          ...      ...         ...   ...       ...     ...            ...   
575491       839  Esteban        Ocon  2024       NaN    1132             \N   
575492       815   Sergio       Pérez  2024       NaN    1132             \N   
575493       855   Guanyu        Zhou  2024       NaN    1132             \N   
575494       847   George     Russell  2024       NaN    1132             \N   
575495       842   Pierre       Gasly  2024       NaN    1132             \N   

        positionOrder  fastestLapTime_m

## Data Transformation

### Convert driver tables to format of final_df

We need a consistent format to merge it with final_df, in order to get one table for each team.

In [ ]:
def forename(name):
  return name.split()[0]

In [ ]:
def surname(name):
  return name.split()[1]

In [ ]:
df_ferrari_drivers_sorted = pd.DataFrame(columns=['forename', 'surname'])
df_mclaren_drivers_sorted = pd.DataFrame(columns=['forename', 'surname'])
df_williams_drivers_sorted = pd.DataFrame(columns=['forename', 'surname'])

df_ferrari_drivers_sorted['forename'] = df_ferrari_drivers['Driver'].apply(forename)
df_mclaren_drivers_sorted['forename'] = df_mclaren_drivers['Driver'].apply(forename)
df_williams_drivers_sorted['forename'] = df_williams_drivers['Driver'].apply(forename)

df_ferrari_drivers_sorted['surname'] = df_ferrari_drivers['Driver'].apply(surname)
df_mclaren_drivers_sorted['surname'] = df_mclaren_drivers['Driver'].apply(surname)
df_williams_drivers_sorted['surname'] = df_williams_drivers['Driver'].apply(surname)

### Merge dataframes

In [ ]:
df_ferrari = fastestlaptimes_final_df.merge(df_ferrari_drivers_sorted, on=['forename', 'surname'], how='inner')
df_mclaren = fastestlaptimes_final_df.merge(df_mclaren_drivers_sorted, on=['forename', 'surname'], how='inner')
df_williams = fastestlaptimes_final_df.merge(df_williams_drivers_sorted, on=['forename', 'surname'], how='inner')

In [ ]:
mask_f = df_ferrari['name'] == 'Ferrari'
df_ferrari = pd.DataFrame(data=df_ferrari[mask_f], columns=['driverId', 'forename', 'surname', 'year', 'name', 'raceId', 'fastestLapTime', 'fastestLapTime_ms', 'positionOrder'])

mask_m = df_mclaren['name'] == 'McLaren'
df_mclaren = pd.DataFrame(data=df_mclaren[mask_m], columns=['driverId', 'forename', 'surname', 'year', 'name', 'raceId', 'fastestLapTime', 'fastestLapTime_ms', 'positionOrder'])

mask_w = df_williams['name'] == 'Williams'
df_williams = pd.DataFrame(data=df_williams[mask_w], columns=['driverId', 'forename', 'surname', 'year', 'name', 'raceId', 'fastestLapTime', 'fastestLapTime_ms', 'positionOrder'])

In [ ]:
df_ferrari.drop_duplicates(subset=['raceId', 'driverId'], keep='first', inplace=True)
df_mclaren.drop_duplicates(subset=['raceId', 'driverId'], keep='first', inplace=True)
df_williams.drop_duplicates(subset=['raceId', 'driverId'], keep='first', inplace=True)

df_ferrari.reset_index(drop=True, inplace=True)
df_mclaren.reset_index(drop=True, inplace=True)
df_williams.reset_index(drop=True, inplace=True)

### Filter for Monza and Monaco

We first prepared seperate dataframes, before realizing, it would be easier to visulise one dataframe, containing all data.

In [ ]:
df_ferrari_monaco = df_ferrari.merge(df_monaco, on='raceId', how='inner')
df_mclaren_monaco = df_mclaren.merge(df_monaco, on='raceId', how='inner')
df_williams_monaco = df_williams.merge(df_monaco, on='raceId', how='inner')

df_ferrari_monza = df_ferrari.merge(df_monza, on='raceId', how='inner')
df_mclaren_monza = df_mclaren.merge(df_monza, on='raceId', how='inner')
df_williams_monza = df_williams.merge(df_monza, on='raceId', how='inner')

In [ ]:
df_monaco_merged = pd.concat([df_ferrari_monaco, df_mclaren_monaco, df_williams_monaco])

df_monza_merged = pd.concat([df_ferrari_monza, df_mclaren_monza, df_williams_monza])

df_monaco_merged.drop(columns=['year_y'], inplace=True)
df_monaco_merged.rename(columns={'year_x': 'year'}, inplace=True)

df_monza_merged.drop(columns=['year_y'], inplace=True)
df_monza_merged.rename(columns={'year_x': 'year'}, inplace=True)

print("Monaco Data:")
print(df_monaco_merged)

print("\nMonza Data:")
df_monza_merged

Monaco Data:
    driverId   forename     surname  year      name  raceId fastestLapTime  \
0         56      Eddie      Irvine  1996   Ferrari     229       1:26.120   
1         30    Michael  Schumacher  1997   Ferrari     211       1:53.315   
2         56      Eddie      Irvine  1997   Ferrari     211       1:54.202   
3         56      Eddie      Irvine  1998   Ferrari     196       1:24.457   
4         30    Michael  Schumacher  1998   Ferrari     196       1:23.189   
..       ...        ...         ...   ...       ...     ...            ...   
28       849   Nicholas      Latifi  2021  Williams    1056       1:15.573   
29       849   Nicholas      Latifi  2022  Williams    1080       1:18.579   
30       848  Alexander       Albon  2022  Williams    1080       1:18.023   
31       848  Alexander       Albon  2023  Williams    1104       1:16.672   
32       848  Alexander       Albon  2024  Williams    1128       1:17.060   

    fastestLapTime_ms  positionOrder  
0          

driverId   forename     surname  year      name  raceId fastestLapTime  \
0         30    Michael  Schumacher  1996   Ferrari     237       1:26.110   
1         56      Eddie      Irvine  1996   Ferrari     237       1:27.687   
2         30    Michael  Schumacher  1997   Ferrari     219       1:25.863   
3         56      Eddie      Irvine  1997   Ferrari     219       1:25.655   
4         30    Michael  Schumacher  1998   Ferrari     204       1:25.483   
..       ...        ...         ...   ...       ...     ...            ...   
27       847     George     Russell  2020  Williams    1038       1:24.421   
28       847     George     Russell  2021  Williams    1065       1:25.835   
29       849   Nicholas      Latifi  2021  Williams    1065       1:25.953   
30       849   Nicholas      Latifi  2022  Williams    1089       1:26.798   
31       848  Alexander       Albon  2023  Williams    1112       1:26.389   

    fastestLapTime_ms  positionOrder  
0             86110.0              1  
1             87687.0             12  
2             85863.0              6  
3             85655.0              8  
4             85483.0              1  
..                ...            ...  
27            84421.0             14  
28            85835.0              9  
29            85953.0             11  
30            86798.0             15  
31            86389.0              7  

[134 rows x 9 columns]

We also removed all outliers, to simplify the graphics.

In [ ]:
def remove_outliers(df, column):
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    df_no_outliers = df[(df[column] >= lower_bound) & (df[column] <= upper_bound)]

    return df_no_outliers

In [ ]:
df_monaco_no_outliers = remove_outliers(df_monaco_merged, 'fastestLapTime_ms')
df_monza_no_outliers = remove_outliers(df_monza_merged, 'fastestLapTime_ms')

### Filter for Hybrid era

We also needed seperate data frames for the Hybrid era.

In [ ]:
df_ferrari_hybrid = df_ferrari[(df_ferrari['year'] >= 2014) & (df_ferrari['year'] <= 2021)]
df_mclaren_hybrid = df_mclaren[(df_mclaren['year'] >= 2014) & (df_mclaren['year'] <= 2021)]
df_williams_hybrid = df_williams[(df_williams['year'] >= 2014) & (df_williams['year'] <= 2021)]

In [ ]:
df_ferrari_hybrid.reset_index(drop=True, inplace=True)
df_mclaren_hybrid.reset_index(drop=True, inplace=True)
df_williams_hybrid.reset_index(drop=True, inplace=True)

In [ ]:
df_hybrid_merged = pd.concat([df_ferrari_hybrid, df_mclaren_hybrid, df_williams_hybrid])
df_hybrid_merged

driverId  forename    surname  year      name  raceId fastestLapTime  \
0           4  Fernando     Alonso  2014   Ferrari     900       1:33.186   
1           8      Kimi  Räikkönen  2014   Ferrari     900       1:33.210   
2           4  Fernando     Alonso  2014   Ferrari     901       1:44.165   
3           8      Kimi  Räikkönen  2014   Ferrari     901       1:45.129   
4           4  Fernando     Alonso  2014   Ferrari     902       1:39.732   
..        ...       ...        ...   ...       ...     ...            ...   
205       849  Nicholas     Latifi  2021  Williams    1071       1:14.616   
206       849  Nicholas     Latifi  2021  Williams    1072       1:32.751   
207       847    George    Russell  2021  Williams    1072       1:36.130   
208       849  Nicholas     Latifi  2021  Williams    1073       1:29.293   
209       847    George    Russell  2021  Williams    1073       1:30.647   

     fastestLapTime_ms  positionOrder  
0              93186.0              4  
1              93210.0              7  
2             104165.0              4  
3             105129.0             12  
4              99732.0              9  
..                 ...            ...  
205            74616.0             16  
206            92751.0             12  
207            96130.0             19  
208            89293.0             16  
209            90647.0             18  

[770 rows x 9 columns]

# 9. Data Visualisation

## Monaco

In [ ]:
fig = px.line(df_monaco_merged, x='year', y='fastestLapTime_ms', color='name',
              title='Fastest Lap Times in Monaco per Team',
              labels={'fastestLapTime_ms': 'Fastest Lap Time (ms)', 'year': 'Year', 'name': 'Team'},
              markers=True,
              color_discrete_sequence=['red', 'orange', 'blue'])

fig.update_layout(
    xaxis_title_font_size=20,
    yaxis_title_font_size=20,
    xaxis = dict(tickfont = dict(size=16)),
    yaxis = dict(tickfont = dict(size=16)),
    title_font_size=24,
    legend_title_font_size=20,
    legend_font_size=16
)

fig.show()

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
fig = px.scatter(df_monaco_no_outliers, x='fastestLapTime_ms', y='positionOrder', color='name',
                 title='Fastest Lap Times and Finishing Position in Monaco per Team',
                 labels={'fastestLapTime_ms': 'Fastest Lap Time (ms)', 'positionOrder': 'Finishing Position', 'name': 'Team'},
                 color_discrete_sequence=['red', 'orange', 'blue'])

fig.update_layout(
    xaxis_title_font_size=20,
    yaxis_title_font_size=20,
    xaxis = dict(tickfont = dict(size=16)),
    yaxis = dict(tickfont = dict(size=16)),
    title_font_size=24,
    legend_title_font_size=20,
    legend_font_size=16
)

fig.show()

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
fig = px.scatter(df_monaco_no_outliers, x='fastestLapTime_ms', y='year', color='name',
                 title='Fastest Lap Times and Year in Monaco per Team',
                 labels={'fastestLapTime_ms': 'Fastest Lap Time (ms)', 'positionOrder': 'Finishing Position', 'name': 'Team'},
                 color_discrete_sequence=['red', 'orange', 'blue'])

fig.update_layout(
    xaxis_title_font_size=20,
    yaxis_title_font_size=20,
    xaxis = dict(tickfont = dict(size=16)),
    yaxis = dict(tickfont = dict(size=16)),
    title_font_size=24,
    legend_title_font_size=20,
    legend_font_size=16
)

fig.show()

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## Monza

In [ ]:
fig = px.line(df_monza_merged, x='year', y='fastestLapTime_ms', color='name',
              title='Fastest Lap Times in Monza per Team',
              labels={'fastestLapTime_ms': 'Fastest Lap Time (ms)', 'year': 'Year', 'name': 'Team'},
              markers=True,
              color_discrete_sequence=['red', 'orange', 'blue'])

fig.update_layout(
    xaxis_title_font_size=20,
    yaxis_title_font_size=20,
    xaxis = dict(tickfont = dict(size=16)),
    yaxis = dict(tickfont = dict(size=16)),
    title_font_size=24,
    legend_title_font_size=20,
    legend_font_size=16
)

fig.show()

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
fig = px.scatter(df_monza_no_outliers, x='fastestLapTime_ms', y='positionOrder', color='name',
                 title='Fastest Lap Times and Finishing Position in Monza per Team',
                 labels={'fastestLapTime_ms': 'Fastest Lap Time (ms)', 'positionOrder': 'Finishing Position', 'name': 'Team'},
                 color_discrete_sequence=['red', 'orange', 'blue'])

fig.update_layout(
    xaxis_title_font_size=20,
    yaxis_title_font_size=20,
    xaxis = dict(tickfont = dict(size=16)),
    yaxis = dict(tickfont = dict(size=16)),
    title_font_size=24,
    legend_title_font_size=20,
    legend_font_size=16
)

fig.show()

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
fig = px.scatter(df_monza_no_outliers, x='fastestLapTime_ms', y='year', color='name',
                 title='Fastest Lap Times and Year in Monza per Team',
                 labels={'fastestLapTime_ms': 'Fastest Lap Time (ms)', 'positionOrder': 'Finishing Position', 'name': 'Team'},
                 color_discrete_sequence=['red', 'orange', 'blue'])

fig.update_layout(
    xaxis_title_font_size=20,
    yaxis_title_font_size=20,
    xaxis = dict(tickfont = dict(size=16)),
    yaxis = dict(tickfont = dict(size=16)),
    title_font_size=24,
    legend_title_font_size=20,
    legend_font_size=16
)

fig.show()

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## Hybrid Era

### Ferrari

In [ ]:
fig = px.line(df_ferrari_hybrid.groupby('raceId')['positionOrder'].min().reset_index(), x='raceId', y='positionOrder',
              title='Finishing Positions in Hybrid Era for Ferrari',
              labels={'positionOrder': 'Best Finishing Position', 'raceId': 'Race'},
              markers=True,
              color_discrete_sequence=['red'])

fig.update_layout(
    xaxis_title_font_size=20,
    yaxis_title_font_size=20,
    xaxis = dict(tickfont = dict(size=16)),
    yaxis = dict(tickfont = dict(size=16)),
    title_font_size=24
)

fig.show()

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
fig = px.line(df_ferrari_hybrid.groupby('raceId')['fastestLapTime_ms'].min().reset_index(), x='raceId', y='fastestLapTime_ms',
              title='Fastest Laps in Hybrid Era for Ferrari',
              labels={'fastestLapTime_ms': 'Fastest Lap Time', 'raceId': 'Race'},
              markers=True,
              color_discrete_sequence=['red'])

fig.update_layout(
    xaxis_title_font_size=20,
    yaxis_title_font_size=20,
    xaxis = dict(tickfont = dict(size=16)),
    yaxis = dict(tickfont = dict(size=16)),
    title_font_size=24
)

fig.show()

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

### McLaren

In [ ]:
fig = px.line(df_mclaren_hybrid.groupby('raceId')['positionOrder'].min().reset_index(), x='raceId', y='positionOrder',
              title='Finishing Positions in Hybrid Era for McLaren',
              labels={'positionOrder': 'Best Finishing Position', 'raceId': 'Race'},
              markers=True,
              color_discrete_sequence=['orange'])

fig.update_layout(
    xaxis_title_font_size=20,
    yaxis_title_font_size=20,
    xaxis = dict(tickfont = dict(size=16)),
    yaxis = dict(tickfont = dict(size=16)),
    title_font_size=24
)

fig.show()

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
fig = px.line(df_mclaren_hybrid.groupby('raceId')['fastestLapTime_ms'].min().reset_index(), x='raceId', y='fastestLapTime_ms',
              title='Fastest Laps in Hybrid Era for McLaren',
              labels={'fastestLapTime_ms': 'Fastest Lap Time', 'raceId': 'Race'},
              markers=True,
              color_discrete_sequence=['orange'])

fig.update_layout(
    xaxis_title_font_size=20,
    yaxis_title_font_size=20,
    xaxis = dict(tickfont = dict(size=16)),
    yaxis = dict(tickfont = dict(size=16)),
    title_font_size=24
)

fig.show()

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

### Williams

In [ ]:
fig = px.line(df_williams_hybrid.groupby('raceId')['positionOrder'].min().reset_index(), x='raceId', y='positionOrder',
              title='Finishing Positions in Hybrid Era for Williams',
              labels={'positionOrder': 'Best Finishing Position', 'raceId': 'Race'},
              markers=True,
              color_discrete_sequence=['blue'])

fig.update_layout(
    xaxis_title_font_size=20,
    yaxis_title_font_size=20,
    xaxis = dict(tickfont = dict(size=16)),
    yaxis = dict(tickfont = dict(size=16)),
    title_font_size=24
)

fig.show()

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
fig = px.line(df_williams_hybrid.groupby('raceId')['fastestLapTime_ms'].min().reset_index(), x='raceId', y='fastestLapTime_ms',
              title='Fastest Laps in Hybrid Era for Williams',
              labels={'fastestLapTime_ms': 'Fastest Lap Time', 'raceId': 'Race'},
              markers=True,
              color_discrete_sequence=['blue'])

fig.update_layout(
    xaxis_title_font_size=20,
    yaxis_title_font_size=20,
    xaxis = dict(tickfont = dict(size=16)),
    yaxis = dict(tickfont = dict(size=16)),
    title_font_size=24
)

fig.show()

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

# 10. Diskussion

In this project, we looked at Formula 1 lap times from 1996 to 2024, combining data about drivers, races, and lap performances. One important step was to convert lap times into milliseconds. This made it easier to work with the data, especially for calculations and comparisons. However, the format is less intuitiv than the original time format (e.g. '1:38.109'), showcasing both formats in future projects could make the data more user-friendly.

The focus on 1996 to 2024 was meant to highlight modern racing trends, but it leaves out earlier history. Including older data could present the history of the sport, but it would also add challenges like inconsistent formats or missing data.

Something else we should point out, is the need to merge the data. With so much information from different sources, there's always a chance for errors, like mismatched columns or missing data. In order to make the process smoother in the future, consistent data (formats), more comprehensive datasets and less tables would be ideal.

# 11. Fazit

This project used different datasets to create an analysis of the development of Formula 1 cars. We filtered the races by year, merged data from multiple sources, and converted lap times into millisecond to make everything precise and consistent.

The final dataset is a great starting point to explore how drivers performed over the years. It can help answer questions such as which driver is the fastest, or how changes in racing rules affected lap times. In the future, adding more details, like weather conditions or car models, could give even deeper insights.

Overall, this project shows how important it is to handle data carefully and think about the details when analzing sports performance. It's an exciting step toward understanding the evolution of Formula 1 racing.